File: artifacts/api-server/src/routes/exams.ts
import { Router, type IRouter } from "express";
import { eq, and, ilike, or } from "drizzle-orm";
import { db, examsTable } from "@workspace/db";
import {
  ListExamsQueryParams,
  ListExamsResponse,
  GetExamResponse,
  GetUpcomingExamsResponse,
} from "@workspace/api-zod";
import { stripNulls, stripNullsArray } from "../lib/nullstrip";

const router: IRouter = Router();

router.get("/exams/upcoming", async (req, res): Promise<void> => {
  const exams = await db
    .select()
    .from(examsTable)
    .where(
      or(
        eq(examsTable.status, "upcoming"),
        eq(examsTable.status, "open")
      )
    )
    .limit(10);

  res.json(GetUpcomingExamsResponse.parse(stripNullsArray(exams)));
});

router.get("/exams", async (req, res): Promise<void> => {
  const params = ListExamsQueryParams.safeParse(req.query);
  if (!params.success) {
    res.status(400).json({ error: "invalid_params", message: params.error.message });
    return;
  }

  const conditions = [];
  if (params.data.category) {
    conditions.push(eq(examsTable.category, params.data.category));
  }
  if (params.data.educationLevel) {
    conditions.push(eq(examsTable.educationLevel, params.data.educationLevel));
  }
  if (params.data.search) {
    conditions.push(
      or(
        ilike(examsTable.name, `%${params.data.search}%`),
        ilike(examsTable.fullName, `%${params.data.search}%`),
        ilike(examsTable.conductingBody, `%${params.data.search}%`)
      )
    );
  }

  const exams = await db
    .select()
    .from(examsTable)
    .where(conditions.length > 0 ? and(...conditions) : undefined)
    .orderBy(examsTable.createdAt);

  res.json(ListExamsResponse.parse(stripNullsArray(exams)));
});

router.get("/exams/:id", async (req, res): Promise<void> => {
  const raw = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const id = parseInt(raw, 10);

  if (isNaN(id)) {
    res.status(400).json({ error: "invalid_params", message: "Invalid exam ID" });
    return;
  }

  const [exam] = await db
    .select()
    .from(examsTable)
    .where(eq(examsTable.id, id));

  if (!exam) {
    res.status(404).json({ error: "not_found", message: "Exam not found" });
    return;
  }

  res.json(GetExamResponse.parse(stripNulls(exam)));
});

export default router;

File: artifacts/api-server/src/routes/exams.rs
use rocket::serde::json::Json;
use rocket::{get, routes, Route, State};
use rocket::form::FromForm;
use sqlx::PgPool;
use crate::errors::{internal_error, not_found_error, ApiResult};
use crate::models::Exam;

const EXAM_SELECT: &str = r#"
    SELECT id, name, full_name, category, education_level, conducting_body,
        application_start_date, application_end_date, exam_date, result_date,
        application_fee, official_website, logo_url, total_vacancies, state,
        is_national, is_urgent, status, description, eligibility, syllabus,
        syllabus_url, application_procedure,
        COALESCE(required_documents, ARRAY[]::text[]) AS required_documents,
        exam_pattern,
        COALESCE(previous_year_papers, ARRAY[]::text[]) AS previous_year_papers,
        created_at
    FROM exams
"#;

#[derive(FromForm)]
pub struct ExamQuery {
    category: Option<String>,
    #[field(name = "educationLevel")]
    education_level: Option<String>,
    search: Option<String>,
}

#[get("/exams/upcoming")]
pub async fn upcoming_exams(pool: &State<PgPool>) -> ApiResult<Vec<Exam>> {
    let sql = format!(
        "{} WHERE status IN ('upcoming', 'open') ORDER BY created_at LIMIT 10",
        EXAM_SELECT
    );
    let exams = sqlx::query_as::<_, Exam>(&sql)
        .fetch_all(pool.inner())
        .await
        .map_err(internal_error)?;
    Ok(Json(exams))
}

#[get("/exams?<q..>")]
pub async fn list_exams(pool: &State<PgPool>, q: ExamQuery) -> ApiResult<Vec<Exam>> {
    let search_pattern = q.search.as_deref().map(|s| format!("%{}%", s));
    let sql = format!(
        r#"{} WHERE ($1::text IS NULL OR category = $1)
           AND ($2::text IS NULL OR education_level = $2)
           AND ($3::text IS NULL OR (name ILIKE $3 OR full_name ILIKE $3 OR conducting_body ILIKE $3))
           ORDER BY created_at"#,
        EXAM_SELECT
    );
    let exams = sqlx::query_as::<_, Exam>(&sql)
        .bind(q.category)
        .bind(q.education_level)
        .bind(search_pattern)
        .fetch_all(pool.inner())
        .await
        .map_err(internal_error)?;
    Ok(Json(exams))
}

#[get("/exams/<id>")]
pub async fn get_exam(pool: &State<PgPool>, id: i32) -> ApiResult<Exam> {
    let sql = format!("{} WHERE id = $1", EXAM_SELECT);
    let exam = sqlx::query_as::<_, Exam>(&sql)
        .bind(id)
        .fetch_optional(pool.inner())
        .await
        .map_err(internal_error)?
        .ok_or_else(|| not_found_error("Exam not found"))?;
    Ok(Json(exam))
}

pub fn routes() -> Vec<Route> {
    routes![upcoming_exams, list_exams, get_exam]
}
